www/docusaurus.preferredTheme.js

```diff
@@ -14,6 +16,46 @@ if (ExecutionEnvironment.canUseDOM) {
     return lightTheme;
   };
 
+  const observer = new MutationObserver(onThemeOrClassChanged);
+  const rootElement = document.documentElement;
+  const themeAttribute = "data-theme";
+
+  observer.observe(rootElement, {
+    attributeFilter: [themeAttribute, "class"],
+    subtree: false,
+  });
+
+  function makeElementVisible(selector) {
+    const codeBlocks = document.querySelectorAll(selector);
+    if(codeBlocks) {
+      codeBlocks.forEach((el) => {
+        el.style.display = 'block';
+      })
+    }
+  }
+
+  function makeElementInvisible(selector) {
+    const codeBlocks = document.querySelectorAll(selector);
+    if(codeBlocks) {
+      codeBlocks.forEach((el) => {
+        el.style.display = 'none';
+      })
+    }
+  }
+
+  function onThemeOrClassChanged() {
+    const theme = htmlElement?.getAttribute('data-theme');
+    if(theme === darkTheme) {
+      makeElementVisible(DARK_THEME_SELECTOR);
+      makeElementInvisible(LIGHT_THEME_SELECTOR);
+    }
+
+    if(theme === lightTheme) {
+      makeElementVisible(LIGHT_THEME_SELECTOR);
+      makeElementInvisible(DARK_THEME_SELECTOR);
+    }
+  }
```

Comment:
Could you please explain how come this is needed?

Reply:
Yes, because the shiki-twoslash support is not as good as for the previous version, so we need to handle theme changing in this way